### PR TITLE
Meson Build System Merge

### PR DESCRIPTION
**Pull request by DMJC**
Implemented Meson Build script. Currently only tested on linux

### CODE REVIEW

**Review by DMJC — COMMENTED** [review]
Meson.build is the only file that needs to be merged initially. In my repository I've merged all the patches that were in the patch folder. Otherwise you can merge my whole repository which will clean out most of the auto-tools configs.

**Comment by DMJC** [comment]
I'm expecting this PR to get knocked back, if you have guidance on how you want this merged I'll try to work with you on it.

**Comment by EionRobb** [comment]
Do you think it would be possible to bring through the original hg commit history from the "missing files" in the pidgin source?

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#PACKAGE_URL = 'https://github.com/DMJC/purple-facebook'
+
+# Configure the package
+#pkgconf = configuration_data()
+#pkgconf.set('PACKAGE_VERSION', PACKAGE_VERSION)
+#pkgconf.set('PACKAGE_URL', PACKAGE_URL)
+
+compiler_options = []
+compiler_options += ['-DPACKAGE_VERSION="0.9.7"']
+compiler_options += ['-DPACKAGE_URL="https://github.com/DMJC/purple-facebook"']
```
Comment (EionRobb) [inline meson.build:14]: Will the url need to be updated after the pr is merged, or would you change as part of this pr?

**Review by EionRobb — APPROVED** [review]
Looks good 🙂 

Have you tried building a Windows dll, out of interest?

File: meson.build (hunk 1 above)
Comment (DMJC) [reply meson.build:14]: It'll change as part of the PR, but it can be edited during the PR afaik. I wasn't sure if the project was maintained or not.

Comment (DMJC) [reply meson.build:14]: I've done a change to the debian files in my repository, the changes now build a working debian package using the Meson build system.

Comment (DMJC) [reply meson.build]: I've updated the URL in the file to have the dequis repository.

**Comment by DMJC** [comment]
Is there anyone with write access who can merge this?

**Comment by grimmy** [comment]
I'll try to take a look tonight.

**Comment by DMJC** [comment]
I have also implemented 2FA for Facebook.

File: README [hunk 2]
```diff
@@ -4,4 +4,16 @@ this plugin is back-ported for purple2.
 
 This project is not affiliated with Facebook, Inc.
 
+This has been forked from https://github.com/dequis/purple-facebook/ - Thanks Dequis for Maintaining this project for so long.
```
Comment (jaymzh) [inline README:7]: This probably shouldn't be merged with this here, it'd be very confusing.

**Review by jaymzh — COMMENTED** [review]
The patches were there, I think, so that it was easy to pull in the upstream bitlbee repos, as those were all patches on top. I think merging them will prevent that....

**Comment by DMJC** [comment]
You can't merge in bitlbee directly, the libpurple and bitlbee apis have
clear differences that prevent merging. I'll clean up my readme.

**Comment by jaymzh** [comment]
> You can't merge in bitlbee directly, the libpurple and bitlbee apis have clear differences that prevent merging. I'll clean up my readme.

I may be confusing different things here, so bear with me, but my understanding is the reason there's a patches directory is because it's checked out from mercurial and patched to work with libpurple. It's described a bit here:

https://github.com/dequis/purple-facebook/issues/516

**Comment by DMJC** [comment]
I've just looked at what it's doing. It's actually using Mercurial to pull code from the Pidgin repository. It's not downloading anything from Bitlbee. The URL: https://keep.imfreedom.org/pidgin/pidgin is the pidgin main repo.

**Comment by DMJC** [comment]
So if What I'm reading is correct, then we should be finding the last version ever of the libpurple mercurial tree before they removed the plugin. Check that out as a hg repository and convert it to git, upload to github to preserve commit history, then we need to delete everything that we don't use, and then we have to go through our patches and verify that they apply, and commit each patch so we have a new history of changes. This is going to take a while.

**Comment by DMJC** [comment]
The biggest problem is there are over 800 authors entries that need to be mapped from Mercurial to GIT. This is a mammoth task.

**Comment by DMJC** [comment]
The revision of the code that the facebook plugin checks out/builds against is: https://keep.imfreedom.org/pidgin/pidgin/rev/9ff9acf9fa14 if you look in the update.sh file in the purple-facebook project, it calls a MANIFEST file and a VERSION file.  The MANIFEST is the files it grabs from Mercurial, the VERSION is the revision it checks out, and there is also a MANIFEST_VOIDS which just creates empty files in includes file the project. Most of the pidgin project code/files is not needed. I need some advice on what we should clean up/keep in terms of history.

**Comment by jaymzh** [comment]
I'm very confused. These questions are probably dumb, but...

Why do we need a fork of pidgin/purple? Why can we not just check out the latest version of pidgin from their authoritative mercurial repo, and just build against that? Why are we "choosing" history to convert to git? I feel like I'm missing some fundamental piece here.

**Comment by DMJC** [comment]
> I'm very confused. These questions are probably dumb, but...

This was written by Dequis the project lead: 
"The whole "check out from mercurial and patch it for libpurple 2.x" thing that this repo is doing is confusing and discourages contributions.

A while ago I hired an intern to help here, and there is a branch called "libpurple-integration" (i didn't pick the name) with steps towards that which merges the relevant hg history of the upstream plugin into this repo.

Sadly it has reached the point where it's my responsibility to review these changes and integrate them, and I can't.

I need someone with maintainer mindset to do the rest of this thing.

Steps forward would be:

    review the commit history
    do any required history rewrites to clean things up
    turn the "patches" directory into real commits (ideally maintaining author metadata)
    modify the build system to not do patching anymore"

Basically I merged everything into git and dumped it to github, now it's easy to checkout and build the project so we can hopefully progress it. The number 1 thing we need to focus on right now is getting end to end encryption working. After that I'd like to get a Windows dll built.

**Comment by jaymzh** [comment]
Yeah I once understood this but I don't anymore. "patch it for libpurpe-2.x" - I'm confused, because there's a [release-2.x.y](https://keep.imfreedom.org/pidgin/pidgin/shortlog/release-2.x.y) branch already. Why are we patching pidgin for pidgin?

**Comment by DMJC** [comment]
We don't anymore, we just keep the entire source tree for the plugin in my
github. I don't follow the history of the plugin, it's essentially an
abandoned project.

**Comment by jaymzh** [comment]
What? I don't mean to be dense here sorry.

This is the plug-in. That we're working on. What history are we talking about? There's pidgin 2, and this, is there something else?
